src/main/java/ch/njol/skript/expressions/ExprEnchantments.java

```diff
@@ -109,36 +124,54 @@ public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
 		
 		switch (mode) {
 			case ADD:
-				for (ItemType item : source)
-					item.addEnchantments(enchants);
+				for (ItemType item : source) {
+					if (isStored)
+						item.addStoredEnchantments(enchants);
+					else
+						item.addEnchantments(enchants);
+				}
 				break;
 			case REMOVE:
 			case REMOVE_ALL:
 				for (ItemType item : source) {
 					ItemMeta meta = item.getItemMeta();
+					EnchantmentStorageMeta storageMeta = item.getEnchantmentStorageMeta();
 					assert meta != null;
 					for (EnchantmentType enchant : enchants) {
 						Enchantment ench = enchant.getType();
 						assert ench != null;
-						if (enchant.getInternalLevel() == -1
-								|| meta.getEnchantLevel(ench) == enchant.getLevel()) {
-							// Remove directly from meta since it's more efficient on this case
-							meta.removeEnchant(ench);
+						if (isStored && storageMeta != null) {
+							if (enchant.getInternalLevel() == -1 || storageMeta.getStoredEnchantLevel(ench) == enchant.getLevel()) {
+								storageMeta.removeStoredEnchant(ench);
+							}
+						}
+						else if (enchant.getInternalLevel() == -1 || meta.getEnchantLevel(ench) == enchant.getLevel()) {
+								// Remove directly from meta since it's more efficient on this case
+								meta.removeEnchant(ench);
```

Comment:
```suggestion
						if (isStored && storageMeta != null) {
							if (storageMeta.getStoredEnchantLevel(ench) == enchant.getLevel()) {
								storageMeta.removeStoredEnchant(ench);
							}
						}
						else if (meta.getEnchantLevel(ench) == enchant.getLevel()) {
								// Remove directly from meta since it's more efficient on this case
								meta.removeEnchant(ench);
```